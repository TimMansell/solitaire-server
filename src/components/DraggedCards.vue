<template>
  <div class="cloned-card" :style="clonedCardStyles">
    <Card
      v-for="(card, index) in clonedCards.cards"
      :key="index"
      :id="card.id"
      :value="card.value"
      :suit="card.suit"
      :revealed="card.revealed"
      :visible="card.visible"
      :clickable="false"
    />
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import { throttle } from 'throttle-debounce';
import Card from '@/components/Card.vue';

export default {
  name: 'DraggedCards',
  components: {
    Card,
  },
  data() {
    return {
      clonedCardStyles: {},
    };
  },
  computed: {
    ...mapGetters(['clonedCards']),
    cardOffset() {
      const { cardWidth } = this.clonedCards;

      return cardWidth / 2;
    },
  },
  mounted() {
    window.addEventListener('dragover', throttle(60, false, this.setStyles));
    window.addEventListener('dragend', this.clearCloneCards);
  },
  destroyed() {
    window.removeEventListener('dragover', this.setStyles);
    window.removeEventListener('dragend', this.clearCloneCards);
  },
  methods: {
    ...mapActions(['clearCloneCards']),
    setStyles(e) {
      // console.log({ e });
      const { x, y } = e;
      const { cardOffset } = this;

      this.clonedCardStyles = {
        top: `${y}px`,
        left: `${x - cardOffset}px`,
      };
    },
  },
};
</script>

<style scoped lang="scss">
.cloned-card {
  position: absolute;
  width: 150px;
}
</style>
