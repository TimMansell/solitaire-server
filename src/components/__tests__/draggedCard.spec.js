import { shallowMount } from '@vue/test-utils';
import DraggedCards from '@/components/DraggedCards.vue';

const defaultProps = {
  width: 100,
};

const computed = {
  draggedCards: () => [
    {
      id: 0,
      value: 'A',
      order: 1,
      suit: '♠',
      visible: true,
    },
    {
      id: 1,
      value: 'A',
      order: 1,
      suit: '♦',
      visible: true,
    },
  ],
};

describe('DraggedCards.vue', () => {
  it('matches visible snapshot', () => {
    const wrapper = shallowMount(DraggedCards, {
      propsData: {
        ...defaultProps,
      },
      computed,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have correct props', () => {
    const wrapper = shallowMount(DraggedCards, {
      propsData: {
        ...defaultProps,
      },
      computed,
    });

    expect(wrapper.props().width).toBe(defaultProps.width);
  });

  it('should calculate card offset correctly', () => {
    const wrapper = shallowMount(DraggedCards, {
      propsData: {
        ...defaultProps,
      },
      computed,
    });

    expect(wrapper.vm.cardOffset).toBe(50);
  });

  it.only('should calculate card position correctly', async () => {
    const wrapper = shallowMount(DraggedCards, {
      propsData: {
        ...defaultProps,
      },
      computed,
    });

    await wrapper.setData({ x: 50, y: 100 });

    expect(wrapper.vm.cardPosition).toEqual({ leftOffset: 0, topOffset: 80 });
  });
});
