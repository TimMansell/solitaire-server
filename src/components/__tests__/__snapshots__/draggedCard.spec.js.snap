// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggedCards.vue matches visible snapshot 1`] = `
<div data-test="dragged-cards-container" class="dragged-cards" style="top: 0px; left: -50px;">
  <div data-test="dragged-cards" style="width: 100px;">
    <card-stub id="dragged-0" value="A" suit="♠" visible="true"></card-stub>
    <card-stub id="dragged-1" value="A" suit="♦" visible="true"></card-stub>
  </div>
</div>
`;
